backend/src/impl/db_utils/dataset_db_utils.py

```diff
@@ -1,42 +1,125 @@
-from typing import Any, Optional
+from __future__ import annotations
 
-from bson import ObjectId
-from explainaboard_web.impl.db_utils.db_utils import DBUtils
+import itertools
+import json
+from datetime import datetime, timedelta
+from typing import Optional
+
+import marisa_trie
+from explainaboard.utils.cache_api import cache_online_file
+from explainaboard.utils.typing_utils import unwrap
 from explainaboard_web.models import DatasetMetadata, DatasetsReturn
 
 
+class DatasetDB:
+    def __init__(self, data: dict):
+        """
+        Create a dataset DB from a list of dictionaries as specified by the jsonl
+        format in DataLab
+        """
+        self.name_dict: dict[str, list[int]] = {}
+        self.task_dict: dict[str, list[int]] = {}
+        self.id_dict: dict[str, int] = {}
+        self.metadatas: list[DatasetMetadata] = []
+        for dataset_name, v_dataset in data.items():
+            for sub_dataset, v_sub in v_dataset["sub_datasets"].items():
+                i = len(self.metadatas)
```

Comment:
It'd be clearer if we could replace `i` with `index` or `idx` or `_id` (the last name seems to be used extensively in the rest of the code) and perhaps add a short explanation of its purpose?